### PR TITLE
Unfuck barricades, use new window code to make them much better

### PR DESCRIPTION
**Pull request by ghost**
Long overdue. Barricades have not been touched ever since they were introduced and hotfixed during Meteors devleopment. While they were created as a straight upgrade to police tape with a pre-existing barricade item, they never really stood out and failed even at the job they were intended for (fortifying hallways and existing window bays, preferably along with keeping the station in about decent shape despite explody explosions)

But now, after having worked on windows and other things, I decided to revisit this, hopefully for the better : 
- Barricade kits now have a new improved sprite which is no longer a copy-pasted plank sprite
  barricade_kit
- You can now create your very own wooden fort with the inclusion of directional barricades to go along full-tile barricades. They have much less health than full-tile barricades (100 for directional, 500 for full, 75 for window/airlock)
- All barricade-related code was moved from deployable (where it didn't belong, given that it's a structure and deployable is under machine) to a new barricade.dm file. A .dmi was also added for quick sprite reference
- Major code unfucking and refactor. All of the code for directional barricades was taken from windows. Note that directional barricades less air pass and aren't opaque (the first is a decision from the fact barricades would be extremely underwhelming and they look solid enough to retain air, the second is due to serious problem with opacity on border structures. Also it inherits windows pretty much fully on pass flags, air doesn't pass)
- Some problems with barricades have been fixed, notably being able to instantly destroy them by spamming attack, and hitting windows/opening airlocks when trying to barricade them (now uses preattack with a proper return)
- Barricades can be deconstructed with a crowbar. It takes ten seconds due to the fact that it instantly kills an otherwise sturdy item (100, or even 500 health is no joke)
- ex_act 1 is needed to kill barricades. Full-tile barricades are extremely beefy and can resist up to five hits

Full showcase with a bit of fluff : 

fort_meteor

Discussion goes below

### CODE REVIEW

**Comment by PJB3005** [comment]
So instead of making these subtype of windows, you copy pasted window code?

top kek

**Comment by dylanstrategie** [comment]
Making them inherit windows would be counterproductive as I would have to override a lot of procs. I would rather copy some procs that fit and tweak a few things

**Comment by PJB3005** [comment]
Give some examples of procs you'd have to overwrite

**Comment by dylanstrategie** [comment]
Procs like Destroy(), healthcheck(), examine(), attackby() since deconstruction is extremely different and in the future potentially other procs since the plan was to make the barricades opaque (but due to Dreammaker hardcoding, opaque border structures make the whole tile opaque like it was a wall, which looks atrocious)

**Comment by PJB3005** [comment]
say window code has 20 procs, and you need to override 10 to make them childs, which is better, overriding 10 or copy pasting 20?

**Comment by dylanstrategie** [comment]
Copy-pasting makes the code much clearer and avoids serious compatibility problems down the line

Barricades seldom work here, it's just a question about how the code should be factored

File: code/game/objects/structures/barricade.dm [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+	..(loc)
+	flags |= ON_BORDER
+	update_icon()
+
+/obj/structure/barricade/examine(mob/user)
+
+	..()
+	//Switch most likely can't take inequalities, so here's that if block
+	if(health >= initial(health)) //Sanity
```
Comment (ComicIronic) [inline code/game/objects/structures/barricade.dm]: Use maxhealth

File: code/game/objects/structures/barricade.dm [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+		health -= I.throwforce
+		healthcheck()
+		visible_message("<span class='danger'>\The [I] slams into \the [src].</span>")
+
+/obj/structure/barricade/attack_hand(mob/user as mob)
+
+	if(M_HULK in user.mutations)
+		user.say(pick(";RAAAAAAAARGH!", ";HNNNNNNNNNGGGGGGH!", ";GWAAAAAAAARRRHHH!", "NNNNNNNNGGGGGGGGHH!", ";AAAAAAARRRGH!"))
+		user.visible_message("<span class='danger'>[user] smashes \the [src]!</span>")
+		health -= 25
```
Comment (ComicIronic) [inline code/game/objects/structures/barricade.dm]: Bad, bad copypaste. This isn't balanced to health.

File: code/game/objects/structures/barricade.dm [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+	var/mob/living/simple_animal/M = user
+	if(M.melee_damage_upper <= 0)
+		return
+	attack_generic(M, M.melee_damage_upper)
+
+/obj/structure/barricade/attack_slime(mob/user as mob)
+
+	if(!isslimeadult(user))
+		return
+	attack_generic(user, rand(10, 15))
```
Comment (ComicIronic) [inline code/game/objects/structures/barricade.dm]: Why these values?

**Comment by MrStonedOne** [comment]
>Copy-pasting makes the code much clearer

kek.

**Comment by d3athrow** [comment]
>Copy-pasting makes the code much clearer and avoids serious compatibility problems down the line

HOLY FUCKING SHIT HOW DENSE ARE YOU
WRITE ONCE ITS DONE, IF A CHANGE HAS TO BE MADE BY SOMEBODY OTHER THAN YOU THEY ONLY HAVE TO CHANGE THE PARENT INSTEAD OF HUNTING DOWN EVERY LITTLE OVERRIDDEN COPY PASTE FIESTA YOU THOUGHT WAS BRILLIANT

**Comment by dylanstrategie** [comment]
But barricades have different functions than windows

**Comment by ComicIronic** [comment]
They look pretty fucking similar to me m80.

**Comment by PJB3005** [comment]
@dylanstrategie, in that they decon differently and some flavour text is different, top kek

**Comment by wwjnc** [comment]
>Copy-pasting makes the code much clearer

https://niceme.me

File: code/game/objects/structures/barricade.dm (hunk 3 above)
Comment (dylanstrategie) [inline code/game/objects/structures/barricade.dm]: What values do you want exactly for that and hulk ?

File: code/game/objects/structures/barricade.dm (hunk 1 above)
Comment (dylanstrategie) [inline code/game/objects/structures/barricade.dm]: Will do

**Comment by Miauw62** [comment]
2015-07-08 19_51_10-unfuck barricades use new window code to make them much better by dylanstrategi

no amount of meme arrows is sufficient

**Comment by RemieRichards** [comment]
You know you've fucked up when you've got us knocking down your door.

**Comment by PJB3005** [comment]
I have a feeling dylan actually became worse at coding than he used to be

**Comment by dylanstrategie** [comment]
You're all getting worked up over a few lines, how hysterical

Anyways, the PR itself is ready. All tested and ready

**Comment by PJB3005** [comment]
\> A few lines

Dylan confirmed unable to count

**Comment by RemieRichards** [comment]
>ready

now THAT is hysterical.

**Comment by ComicIronic** [comment]
TG GO HOME

**Comment by Aranclanos** [comment]
TG  MASTERRACE

**Comment by Cheridan** [comment]
we came here to laugh at you

**Comment by optimumtact** [comment]
move the common procs into a higher level subtype

**Comment by Iamgoofball** [comment]
As the official currently active worstcoder in all of SS13, I'd like to state that you fucked up and I'm dedicating my chere of supa ideas to you.

startoad

**Comment by xxalpha** [comment]
You should learn to listen to the people with more experience than you.

**Comment by Iamgoofball** [comment]
such as me

**Comment by Miauw62** [comment]
>tfw tg raids a vg pr

**Comment by PJB3005** [comment]
TIME TO RAID A TG PR

**Comment by SkowronX** [comment]
Feels more like a friendly get-together to laugh at something silly.

**Comment by optimumtact** [comment]
I'm sure OP was well intentioned

**Comment by Aranclanos** [comment]
yall bunch of bullies

**Pull request closed without merging**